code/Model/SiteTree.php

```diff
@@ -1338,61 +1338,102 @@ public function collateDescendants($condition, &$collator)
     }
 
     /**
-     * Return the title, description, keywords and language metatags.
-     *
-     * @todo Move <title> tag in separate getter for easier customization and more obvious usage
+     * Return the title, description, keywords and language metatags in a keyed array
      *
-     * @param bool $includeTitle Show default <title>-tag, set to false for custom templating
-     * @return string The XHTML metatags
+     * @return array array of metatag properties ready to be turned in to html tags with HTML::create_tag()
      */
-    public function MetaTags($includeTitle = true)
+    public function MetaTagArray()
     {
-        $tags = array();
-        if ($includeTitle && strtolower($includeTitle) != 'false') {
-            $tags[] = HTML::createTag('title', array(), $this->obj('Title')->forTemplate());
-        }
+        $tags = [];
+
+        $tags['title'] = [
+            'tag' => 'title',
+            'attributes' => [],
+            'content' => $this->obj('Title')->forTemplate()
+        ];
 
         $generator = trim(Config::inst()->get(self::class, 'meta_generator'));
         if (!empty($generator)) {
-            $tags[] = HTML::createTag('meta', array(
-                'name' => 'generator',
-                'content' => $generator,
-            ));
+            $tags['generator'] = [
+                'tag' => 'meta',
+                'attributes' => [
+                    'name' => 'generator',
+                    'content' => $generator,
+                ],
+            ];
         }
 
         $charset = ContentNegotiator::config()->uninherited('encoding');
-        $tags[] = HTML::createTag('meta', array(
-            'http-equiv' => 'Content-Type',
-            'content' => 'text/html; charset=' . $charset,
-        ));
+        $tags['contentType'] = [
+            'tag' => 'meta',
+            'attributes' => [
+                'http-equiv' => 'Content-Type',
+                'content' => 'text/html; charset=' . $charset,
+            ],
+        ];
         if ($this->MetaDescription) {
-            $tags[] = HTML::createTag('meta', array(
-                'name' => 'description',
-                'content' => $this->MetaDescription,
-            ));
+            $tags['description'] = [
+                'tag' => 'meta',
+                'attributes' => [
+                    'name' => 'description',
+                    'content' => $this->MetaDescription,
+                ],
+            ];
         }
 
         if (Permission::check('CMS_ACCESS_CMSMain')
             && $this->ID > 0
         ) {
-            $tags[] = HTML::createTag('meta', array(
-                'name' => 'x-page-id',
-                'content' => $this->obj('ID')->forTemplate(),
-            ));
-            $tags[] = HTML::createTag('meta', array(
-                'name' => 'x-cms-edit-link',
-                'content' => $this->obj('CMSEditLink')->forTemplate(),
```

Comment:
Side note: `HTML::createTag` calls `Convert::raw2att()` on each attribute so aren't all these calls to `->forTemplate()` (apart from the Title field) causing double escaping? Should I change them to call `->getValue()` instead, or skip the `->obj()` call to get the raw values?

Reply:
Currently SiteTree defines $casting for MetaTags to HTMLFragment, which doesn't escape when rendered in the template, so it shouldn't be double escaping unless people put `$MetaTags.XML` in their template rather than `$MetaTags`

It would probably be better to have a template include that loops through your new array method and renders the meta tags in the template instead of doing it all here - would you agree?

Reply:
It's not the whole `$MetaTags` object that I'm talking about here, just some attributes on individual tags e.g. `$tags['cmsEditLink']['attributes']['content']` is set to `$this->obj('CMSEditLink')->forTemplate()` which I assume is an escaped value, which is then passed to `Convert::raw2att()`. It's actually only the `cmsEditLink` and `pageId` values that get this treatment, and I would guess no escaping is needed on those values which is why the double escaping hasn't surfaced as a problem, since escaping doesn't change the values. So I'd say it's probably harmless but may as well be corrected.

Reply:
Ah I see. Yeah - if we _were_ to move this into a template then the attribute escaping can be done explicitly in the template as well perhaps.

In the case of CMSEditLink, `forTemplate()` would cause whatever casting is defined for SiteTree which is Text, and that model does escape by default (`DBField::forTemplate`) so I think you're right, the Convert::raw2att here is causing double escaping and you're also right that it's not surfacing as a problem because of the contents of the existing data we're using =)

Nice to spitball this!